File: go/engine_test.go
package hermetix

// 엔진 로직 검증 - 다른 언어 구현과 동일한 시나리오.

import (
	"testing"
	"time"

	"github.com/shopspring/decimal"
)

type fakeBroker struct {
	created     []CreateOrderRequest
	canceled    []string
	orderStatus OrderStatus
	openOrders  []Order
	environment TradingEnvironment
}

func newFakeBroker() *fakeBroker {
	return &fakeBroker{orderStatus: Submitted, environment: Paper}
}

func (f *fakeBroker) Capabilities() BrokerCapabilities {
	return BrokerCapabilities{
		BrokerID: "fake", Market: "US", Currency: "USD",
		CandleIntervals: map[CandleInterval]bool{Day1: true},
		ClientOrderID:   true, ServerOpenOrders: true,
		Environments: map[TradingEnvironment]bool{Paper: true, Live: true},
	}
}
func (f *fakeBroker) Environment() TradingEnvironment                          { return f.environment }
func (f *fakeBroker) GetQuotes([]string) ([]Quote, error)                      { return nil, nil }
func (f *fakeBroker) GetCandles(string, CandleInterval, int) ([]Candle, error) { return nil, nil }
func (f *fakeBroker) GetCalendar() ([]MarketDay, error)                        { return nil, nil }
func (f *fakeBroker) GetAccount() (Account, error)                             { return Account{}, nil }
func (f *fakeBroker) GetHoldings() ([]Holding, error)                          { return nil, nil }
func (f *fakeBroker) GetBuyingPower() (decimal.Decimal, error)                 { return decimal.Zero, nil }
func (f *fakeBroker) CreateOrder(request CreateOrderRequest) (Order, error) {
	f.created = append(f.created, request)
	return Order{OrderID: "ord_1", Status: Submitted, Symbol: request.Symbol}, nil
}
func (f *fakeBroker) GetOrders() ([]Order, error) { return f.openOrders, nil }
func (f *fakeBroker) GetOrder(orderID string) (Order, error) {
	return Order{OrderID: orderID, Status: f.orderStatus}, nil
}
func (f *fakeBroker) CancelOrder(orderID string) (Order, error) {
	f.canceled = append(f.canceled, orderID)
	return Order{OrderID: orderID, Status: Canceled}, nil
}
func (f *fakeBroker) GetFills() ([]Fill, error) { return nil, nil }

func testCtx(price string, heldQty string) *StrategyContext {
	ctx := &StrategyContext{
		Now: time.Now(), Quotes: map[string]Quote{}, Candles: map[string][]Candle{},
		Holdings: map[string]Holding{}, BuyingPower: decimal.NewFromInt(10000),
	}
	if price != "" {
		p, _ := decimal.NewFromString(price)
		ctx.Quotes["AAPL"] = Quote{Symbol: "AAPL", Price: p, Timestamp: time.Now()}
	}
	if heldQty != "" {
		q, _ := decimal.NewFromString(heldQty)
		ctx.Holdings["AAPL"] = Holding{Symbol: "AAPL", Quantity: q, AvgEntryPrice: decimal.NewFromInt(100)}
	}
	return ctx
}

func TestSellClampedToHoldings(t *testing.T) {
	broker := newFakeBroker()
	executor := NewOrderExecutor(broker, NewBracketMonitor(broker), NewTradingGuard(broker, 5))
	executor.Execute("t", []Signal{SellSignal{Symbol: "AAPL", Quantity: decimal.NewFromInt(10)}}, testCtx("", "3"))
	if broker.created[0].Quantity.String() != "3" {
		t.Fatalf("quantity = %s", broker.created[0].Quantity)
	}
	if broker.created[0].ClientOrderID == "" {
		t.Fatal("clientOrderID should be set")
	}
}

func TestSellSkippedWithoutHoldings(t *testing.T) {
	broker := newFakeBroker()
	executor := NewOrderExecutor(broker, NewBracketMonitor(broker), NewTradingGuard(broker, 5))
	executor.Execute("t", []Signal{SellSignal{Symbol: "AAPL", Quantity: decimal.NewFromInt(1)}}, testCtx("", ""))
	if len(broker.created) != 0 {
		t.Fatal("should not create order")
	}
}

func TestHaltedGuardBlocksSignals(t *testing.T) {
	broker := newFakeBroker()
	guard := NewTradingGuard(broker, 5)
	guard.Halt("test")
	executor := NewOrderExecutor(broker, NewBracketMonitor(broker), guard)
	executor.Execute("t", []Signal{BuySignal{Symbol: "AAPL", Quantity: decimal.NewFromInt(1)}}, testCtx("", ""))
	if len(broker.created) != 0 {
		t.Fatal("should not create order while halted")
	}
}

func TestBracketTakeProfitFlow(t *testing.T) {
	broker := newFakeBroker()
	brackets := NewBracketMonitor(broker)
	executor := NewOrderExecutor(broker, brackets, NewTradingGuard(broker, 5))

	tp, sl := decimal.NewFromInt(310), decimal.NewFromInt(280)
	executor.Execute("t", []Signal{BuySignal{
		Symbol: "AAPL", Quantity: decimal.NewFromInt(2), TakeProfitPrice: &tp, StopLossPrice: &sl,
	}}, testCtx("", ""))
	if brackets.ActiveCount() != 1 {
		t.Fatalf("activeCount = %d", brackets.ActiveCount())
	}

	broker.orderStatus = Filled // 진입 체결
	signals := brackets.Check(testCtx("311", "2"))
	if len(signals) != 1 {
		t.Fatalf("signals = %d", len(signals))
	}
	if signals[0].(SellSignal).Quantity.String() != "2" {
		t.Fatal("wrong quantity")
	}
	if brackets.ActiveCount() != 0 {
		t.Fatal("bracket should be removed")
	}
}

func TestBracketDroppedWhenEntryCanceled(t *testing.T) {
	broker := newFakeBroker()
	brackets := NewBracketMonitor(broker)
	tp := decimal.NewFromInt(310)
	brackets.Register("ord_1", "AAPL", decimal.NewFromInt(2), &tp, nil)
	broker.orderStatus = Canceled
	if signals := brackets.Check(testCtx("999", "")); len(signals) != 0 {
		t.Fatal("should not signal")
	}
	if brackets.ActiveCount() != 0 {
		t.Fatal("bracket should be dropped")
	}
}

func TestGuardHaltsAfterConsecutiveFailures(t *testing.T) {
	broker := newFakeBroker()
	broker.openOrders = []Order{{OrderID: "ord_9", Status: Submitted}}
	guard := NewTradingGuard(broker, 3)
	for i := 0; i < 3; i++ {
		guard.RecordFailure(&BrokerAPIError{500, "", "boom"})
	}
	if !guard.IsHalted() {
		t.Fatal("should be halted")
	}
	if len(broker.canceled) != 1 || broker.canceled[0] != "ord_9" {
		t.Fatalf("canceled = %v", broker.canceled)
	}
}

// ------------------------------------------------------------- 0.6.0: 위험 상한 · 실전 게이트 · 심볼 접두

func dec(s string) *decimal.Decimal {
	v, _ := decimal.NewFromString(s)
	return &v
}

func TestRiskGuardLimits(t *testing.T) {
	guard := NewRiskGuard(dec("1000"), nil)
	if guard.TryReserve("AAPL", decimal.NewFromInt(2), dec("600")) == "" {
		t.Fatal("1200 > 1000 이면 거부")
	}
	if guard.TryReserve("AAPL", decimal.NewFromInt(1), dec("600")) != "" {
		t.Fatal("600 은 통과")
	}
	if guard.TryReserve("AAPL", decimal.NewFromInt(1), nil) == "" {
		t.Fatal("가격 미상은 거부")
	}
	if NewRiskGuard(nil, nil).TryReserve("AAPL", decimal.NewFromInt(1), nil) != "" {
		t.Fatal("상한 없으면 통과")
	}

	now := time.Date(2026, 9, 10, 23, 0, 0, 0, time.UTC)
	daily := NewRiskGuard(nil, dec("1000"))
	daily.now = func() time.Time { return now }
	daily.day = daily.today()
	if daily.TryReserve("AAPL", decimal.NewFromInt(3), dec("300")) != "" {
		t.Fatal("900 통과")
	}
	if daily.TryReserve("AAPL", decimal.NewFromInt(1), dec("300")) == "" {
		t.Fatal("1200 거부")
	}
	now = time.Date(2026, 9, 11, 1, 0, 0, 0, time.UTC)
	if daily.TryReserve("AAPL", decimal.NewFromInt(1), dec("300")) != "" {
		t.Fatal("UTC 자정 경과 후 초기화")
	}
	if daily.DailyTotal().String() != "300" {
		t.Fatalf("dailyTotal = %s", daily.DailyTotal())
	}
}

func TestExecutorSkipsOrdersOverRiskLimit(t *testing.T) {
	broker := newFakeBroker()
	executor := NewOrderExecutor(broker, NewBracketMonitor(broker), NewTradingGuard(broker, 5)).WithRisk(NewRiskGuard(dec("1000"), nil))
	executor.Execute("t", []Signal{
		BuySignal{Symbol: "AAPL", Quantity: decimal.NewFromInt(5)},
		BuySignal{Symbol: "AAPL", Quantity: decimal.NewFromInt(3)},
		BuySignal{Symbol: "AAPL", Quantity: decimal.NewFromInt(10), OrderType: Limit, LimitPrice: dec("50")},
		BuySignal{Symbol: "NOPE", Quantity: decimal.NewFromInt(1)},
	}, testCtx("300", ""))
	if len(broker.created) != 2 || broker.created[0].Quantity.String() != "3" || broker.created[1].Quantity.String() != "10" {
		t.Fatalf("created = %+v", broker.created)
	}
}

type noopStrategy struct{}

func (noopStrategy) Spec() StrategySpec                        { return StrategySpec{Name: "t", Symbols: []string{"AAPL"}} }
func (noopStrategy) Decide(*StrategyContext) ([]Signal, error) { return nil, nil }

func TestLiveGateRequiresExplicitConsent(t *testing.T) {
	broker := newFakeBroker()
	broker.environment = Live
	if n := len(NewStrategyEngine(broker, []Strategy{noopStrategy{}}).Strategies); n != 0 {
		t.Fatalf("LIVE 는 동의 없이 스케줄되면 안 된다: %d", n)
	}
	if n := len(NewStrategyEngineWithOptions(broker, []Strategy{noopStrategy{}}, EngineOptions{LiveTradingEnabled: true}).Strategies); n != 1 {
		t.Fatalf("LIVE + 동의 = 스케줄: %d", n)
	}
	broker.environment = Paper
	if n := len(NewStrategyEngine(broker, []Strategy{noopStrategy{}}).Strategies); n != 1 {
		t.Fatalf("PAPER 는 스케줄: %d", n)
	}
}

func TestContextMatchesSymbolsIgnoringMarketPrefix(t *testing.T) {
	ctx := &StrategyContext{
		Quotes:     map[string]Quote{"KRX:005930": {Symbol: "KRX:005930", Price: decimal.NewFromInt(70000)}},
		Holdings:   map[string]Holding{"005930": {Symbol: "005930", Quantity: decimal.NewFromInt(3)}},
		OpenOrders: []Order{{OrderID: "o1", Status: Submitted, Symbol: "005930"}},
	}
	if q, ok := ctx.Quote("005930"); !ok || q.Price.String() != "70000" {
		t.Fatal("접두 없는 조회가 접두 있는 키에 맞아야 한다")
	}
	if h, ok := ctx.Holding("KRX:005930"); !ok || h.Quantity.String() != "3" {
		t.Fatal("접두 있는 조회가 접두 없는 키에 맞아야 한다")
	}
	if !ctx.HasPosition("KRX:005930") || !ctx.HasOpenOrder("KRX:005930") {
		t.Fatal("HasPosition/HasOpenOrder")
	}
	if _, ok := ctx.Quote("AAPL"); ok {
		t.Fatal("없는 심볼")
	}
}

File: js/tests/engine.test.ts
/** 엔진 로직 검증 - Kotlin/Python 레퍼런스와 동일 시나리오. */
import assert from "node:assert/strict";
import { test } from "node:test";
import type { BrokerClient } from "../src/broker.js";
import { BracketMonitor, OrderExecutor, RiskGuard, StrategyEngine, TradingGuard } from "../src/engine.js";
import { Decimal } from "../src/models.js";
import type { CreateOrderRequest, Order, OrderStatus, TradingEnvironment } from "../src/models.js";
import { StrategyContext, buy, sell } from "../src/strategy.js";

class FakeBroker implements BrokerClient {
  capabilities = {
    brokerId: "fake", market: "US" as const, currency: "USD",
    candleIntervals: new Set(["1d" as const]),
    clientOrderId: true, nativeBracket: false, fractionalShares: false, serverOpenOrders: true,
    environments: new Set<TradingEnvironment>(["PAPER", "LIVE"]),
  };
  environment: TradingEnvironment = "PAPER";
  created: CreateOrderRequest[] = [];
  canceled: string[] = [];
  orderStatus: OrderStatus = "SUBMITTED";
  openOrders: Order[] = [];

  async getQuotes() { return []; }
  async getCandles() { return []; }
  async getCalendar() { return []; }
  async getAccount() { throw new Error("unused"); return undefined as never; }
  async getHoldings() { return []; }
  async getBuyingPower() { return new Decimal(0); }
  async createOrder(request: CreateOrderRequest): Promise<Order> {
    this.created.push(request);
    return { orderId: `ord_${this.created.length}`, status: "SUBMITTED", symbol: request.symbol };
  }
  async getOrders() { return this.openOrders; }
  async getOrder(orderId: string): Promise<Order> { return { orderId, status: this.orderStatus }; }
  async cancelOrder(orderId: string): Promise<Order> {
    this.canceled.push(orderId);
    return { orderId, status: "CANCELED" };
  }
  async getFills() { return []; }
}

const ctx = (opts: { price?: string; heldQty?: string } = {}) =>
  new StrategyContext(
    new Date(),
    new Map(opts.price ? [["AAPL", {
      symbol: "AAPL", price: new Decimal(opts.price), bidPrice: null, askPrice: null,
      volume: 0, change: null, changeRate: null, timestamp: new Date(),
    }]] : []),
    new Map(),
    { accountId: "a", currency: "USD", cash: new Decimal(1), portfolioValue: new Decimal(1), status: "ACTIVE" },
    new Map(opts.heldQty ? [["AAPL", {
      symbol: "AAPL", quantity: new Decimal(opts.heldQty), avgEntryPrice: new Decimal(100),
    }]] : []),
    [],
    new Decimal(10000),
  );

test("매도는 보유 수량으로 클램프", async () => {
  const broker = new FakeBroker();
  const executor = new OrderExecutor(broker, new BracketMonitor(broker), new TradingGuard(broker));
  await executor.execute("t", [sell("AAPL", new Decimal(10))], ctx({ heldQty: "3" }));
  assert.equal(broker.created[0].quantity.toString(), "3");
  assert.ok(broker.created[0].clientOrderId!.startsWith("t-"));
});

test("보유 없으면 매도 스킵", async () => {
  const broker = new FakeBroker();
  const executor = new OrderExecutor(broker, new BracketMonitor(broker), new TradingGuard(broker));
  await executor.execute("t", [sell("AAPL", new Decimal(1))], ctx());
  assert.equal(broker.created.length, 0);
});

test("비상정지 중 시그널 차단", async () => {
  const broker = new FakeBroker();
  const guard = new TradingGuard(broker);
  await guard.halt("test");
  const executor = new OrderExecutor(broker, new BracketMonitor(broker), guard);
  await executor.execute("t", [buy("AAPL", new Decimal(1))], ctx());
  assert.equal(broker.created.length, 0);
});

test("브라켓 익절 플로우", async () => {
  const broker = new FakeBroker();
  const brackets = new BracketMonitor(broker);
  const executor = new OrderExecutor(broker, brackets, new TradingGuard(broker));

  await executor.execute("t", [buy("AAPL", new Decimal(2), {
    takeProfitPrice: new Decimal(310), stopLossPrice: new Decimal(280),
  })], ctx());
  assert.equal(brackets.activeCount, 1);

  broker.orderStatus = "FILLED"; // 진입 체결
  const signals = await brackets.check(ctx({ price: "311", heldQty: "2" }));
  assert.equal(signals.length, 1);
  assert.equal(signals[0].quantity.toString(), "2");
  assert.equal(brackets.activeCount, 0);
});

test("진입 취소 시 브라켓 폐기", async () => {
  const broker = new FakeBroker();
  const brackets = new BracketMonitor(broker);
  brackets.register("ord_1", "AAPL", new Decimal(2), new Decimal(310), null);
  broker.orderStatus = "CANCELED";
  assert.equal((await brackets.check(ctx({ price: "999" }))).length, 0);
  assert.equal(brackets.activeCount, 0);
});

test("연속 실패 시 비상정지 + 미체결 취소", async () => {
  const broker = new FakeBroker();
  broker.openOrders = [{ orderId: "ord_9", status: "SUBMITTED" }];
  const guard = new TradingGuard(broker, 3);
  for (let i = 0; i < 3; i++) await guard.recordFailure(new Error("boom"));
  assert.ok(guard.isHalted);
  assert.deepEqual(broker.canceled, ["ord_9"]);
});


// ------------------------------------------------------------- 0.6.0: 위험 상한 · 실전 게이트 · 심볼 접두

test("RiskGuard: 1건/일일 상한, 가격 미상 거부, UTC 일자 변경 시 초기화", () => {
  const guard = new RiskGuard(new Decimal(1000));
  assert.ok(guard.tryReserve("AAPL", new Decimal(2), new Decimal(600)));
  assert.equal(guard.tryReserve("AAPL", new Decimal(1), new Decimal(600)), null);
  assert.ok(guard.tryReserve("AAPL", new Decimal(1), null));
  assert.equal(new RiskGuard().tryReserve("AAPL", new Decimal(1), null), null);

  let now = new Date("2026-09-10T23:00:00Z");
  const daily = new RiskGuard(null, new Decimal(1000), () => now);
  assert.equal(daily.tryReserve("AAPL", new Decimal(3), new Decimal(300)), null);
  assert.ok(daily.tryReserve("AAPL", new Decimal(1), new Decimal(300)));
  now = new Date("2026-09-11T01:00:00Z");
  assert.equal(daily.tryReserve("AAPL", new Decimal(1), new Decimal(300)), null);
  assert.equal(daily.getDailyTotal().toString(), "300");
});

test("실행기는 금액 상한을 넘는 시그널을 제출하지 않는다", async () => {
  const broker = new FakeBroker();
  const executor = new OrderExecutor(broker, new BracketMonitor(broker), new TradingGuard(broker), new RiskGuard(new Decimal(1000)));
  await executor.execute("t", [
    buy("AAPL", new Decimal(5)),
    buy("AAPL", new Decimal(3)),
    buy("AAPL", new Decimal(10), { orderType: "LIMIT", limitPrice: new Decimal(50) }),
    buy("NOPE", new Decimal(1)),
  ], ctx({ price: "300" }));
  assert.deepEqual(broker.created.map((r) => r.quantity.toString()), ["3", "10"]);
});

test("실전 게이트: LIVE 는 liveTradingEnabled 없이는 스케줄되지 않는다", () => {
  const noop = { spec: { name: "t", symbols: ["AAPL"] }, decide: () => [] };
  const broker = new FakeBroker();
  broker.environment = "LIVE";
  assert.equal(new StrategyEngine(broker, [noop]).strategies.length, 0);
  assert.equal(new StrategyEngine(broker, [noop], 5, { liveTradingEnabled: true }).strategies.length, 1);
  broker.environment = "PAPER";
  assert.equal(new StrategyEngine(broker, [noop]).strategies.length, 1);
});

test("컨텍스트 조회는 시장 접두 유무를 무시한다", () => {
  const c = new StrategyContext(
    new Date(),
    new Map([["KRX:005930", { symbol: "KRX:005930", price: new Decimal(70000), bidPrice: null, askPrice: null, volume: 0, change: null, changeRate: null, timestamp: new Date() }]]),
    new Map(),
    { accountId: "a", currency: "KRW", cash: new Decimal(1), portfolioValue: new Decimal(1), status: "ACTIVE" },
    new Map([["005930", { symbol: "005930", quantity: new Decimal(3), avgEntryPrice: new Decimal(1) }]]),
    [{ orderId: "o1", status: "SUBMITTED", symbol: "005930" }],
    new Decimal(1),
  );
  assert.equal(c.quote("005930")!.price.toString(), "70000");
  assert.equal(c.holding("KRX:005930")!.quantity.toString(), "3");
  assert.equal(c.hasPosition("KRX:005930"), true);
  assert.equal(c.hasOpenOrder("KRX:005930"), true);
  assert.equal(c.quote("AAPL"), undefined);
});
